plonky2/src/gates/gate.rs

```diff
@@ -17,11 +18,38 @@ use crate::plonk::circuit_builder::CircuitBuilder;
 use crate::plonk::vars::{
     EvaluationTargets, EvaluationVars, EvaluationVarsBase, EvaluationVarsBaseBatch,
 };
+use crate::util::serialization::Buffer;
+
+pub enum GateKind {
+    ArithmeticBase,
+    ArithmeticExt,
+    AssertLe,
+    BaseSum,
+    Constant,
+    Exponentiation,
+    Interpolation,
+    LowDegreeInterpolation,
+    MulExt,
+    Noop,
+    PoseidonMds,
+    Poseidon,
+    PublicInput,
+    RandomAccess,
+    ReducingExt,
+    Reducing,
+}
 
 /// A custom gate.
 pub trait Gate<F: RichField + Extendable<D>, const D: usize>: 'static + Send + Sync {
     fn id(&self) -> String;
 
+    fn kind(&self) -> GateKind;
```

Comment:
One drawback of this approach is that new gates cannot be defined in external crates (as can be seen from the failing test). 
Do you really need this `GateKind` enum in your applications, or would using the `id()` string as a proxy also work?

Reply:
Maybe we could just have the serialization function take a vector of supported gates `gates: Vec<Box<dyn Gate<F, D>>>` and have the tag of a gate be its index in the vector. That way we don't need the `GateKind` enum nor the `kind()` function. 
Serialization would be dependent on an agreed-upon set of gates for the application, but I think that's a feature as different applications might require different sets of gates.

Reply:
Good idea

Reply:
Hm good point. It's a bit hacky but I guess we could have `deserialize` be non-static.